File: snippets/php/90-lab/04-links.php
<?php
// The runner places the shared teaching client beside this example.
// ELOPG_PLAN: [{"method":"linkSords","params":{"fromId":{"$expression":"a"},"toIds":[{"$expression":"b"}],"linkZ":{"bset":"1"}}},{"method":"unlinkSords","params":{"fromId":{"$expression":"a"},"toIds":[{"$expression":"b"}],"linkZ":{"bset":"1"}}},{"method":"createSord","params":{"parentId":{"$expression":"str(PARENT)"},"maskId":0,"editInfoZ":{"bset":"1","sordZ":{"bset":{"$expression":"ALL"}}}}},{"method":"checkinSord","params":{"sord":{"$expression":"tpl"},"sordZ":{"bset":{"$expression":"ALL"}},"unlockZ":{"bset":"1"}}},{"method":"checkoutSord","params":{"objId":{"$expression":"a"},"editInfoZ":{"bset":"1","sordZ":{"bset":{"$expression":"ALL"}}}}},{"method":"deleteSord","params":{"parentId":{"$expression":"str(PARENT)"},"objId":{"$expression":"oid"},"deleteOptions":{"deleteFinally":false}}},{"method":"deleteSord","params":{"parentId":{"$expression":"str(PARENT)"},"objId":{"$expression":"oid"},"deleteOptions":{"deleteFinally":true}}}]
require_once __DIR__ . '/EloClient.php';

$ELO_BASE_URL = getenv('ELOPG_ELO_BASE_URL') ?: 'http://localhost:9090/ix-Repository1'; // ELOPG_DEFAULT:base_url
$ELO_USER = getenv('ELOPG_ELO_USER') ?: 'Administrator'; // ELOPG_DEFAULT:user
$ELO_PASS = getenv('ELOPG_ELO_PASSWORD') ?: 'elo'; // ELOPG_DEFAULT:password
$elo = EloClient::connect($ELO_BASE_URL, $ELO_USER, $ELO_PASS);
$result = $elo->call('linkSords', json_decode('{"fromId":{"$expression":"a"},"toIds":[{"$expression":"b"}],"linkZ":{"bset":"1"}}', true, 512, JSON_THROW_ON_ERROR));
echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES), PHP_EOL;
$result = $elo->call('unlinkSords', json_decode('{"fromId":{"$expression":"a"},"toIds":[{"$expression":"b"}],"linkZ":{"bset":"1"}}', true, 512, JSON_THROW_ON_ERROR));
echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES), PHP_EOL;
$result = $elo->call('createSord', json_decode('{"parentId":{"$expression":"str(PARENT)"},"maskId":0,"editInfoZ":{"bset":"1","sordZ":{"bset":{"$expression":"ALL"}}}}', true, 512, JSON_THROW_ON_ERROR));
echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES), PHP_EOL;
$result = $elo->call('checkinSord', json_decode('{"sord":{"$expression":"tpl"},"sordZ":{"bset":{"$expression":"ALL"}},"unlockZ":{"bset":"1"}}', true, 512, JSON_THROW_ON_ERROR));
echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES), PHP_EOL;
$result = $elo->call('checkoutSord', json_decode('{"objId":{"$expression":"a"},"editInfoZ":{"bset":"1","sordZ":{"bset":{"$expression":"ALL"}}}}', true, 512, JSON_THROW_ON_ERROR));
echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES), PHP_EOL;
$result = $elo->call('deleteSord', json_decode('{"parentId":{"$expression":"str(PARENT)"},"objId":{"$expression":"oid"},"deleteOptions":{"deleteFinally":false}}', true, 512, JSON_THROW_ON_ERROR));
echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES), PHP_EOL;
$result = $elo->call('deleteSord', json_decode('{"parentId":{"$expression":"str(PARENT)"},"objId":{"$expression":"oid"},"deleteOptions":{"deleteFinally":true}}', true, 512, JSON_THROW_ON_ERROR));
echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES), PHP_EOL;

File: snippets/java/90-lab/04-links.java
// The runner compiles this file with shared/java/EloClient.java.
// ELOPG_PLAN: [{"method":"linkSords","params":{"fromId":{"$expression":"a"},"toIds":[{"$expression":"b"}],"linkZ":{"bset":"1"}}},{"method":"unlinkSords","params":{"fromId":{"$expression":"a"},"toIds":[{"$expression":"b"}],"linkZ":{"bset":"1"}}},{"method":"createSord","params":{"parentId":{"$expression":"str(PARENT)"},"maskId":0,"editInfoZ":{"bset":"1","sordZ":{"bset":{"$expression":"ALL"}}}}},{"method":"checkinSord","params":{"sord":{"$expression":"tpl"},"sordZ":{"bset":{"$expression":"ALL"}},"unlockZ":{"bset":"1"}}},{"method":"checkoutSord","params":{"objId":{"$expression":"a"},"editInfoZ":{"bset":"1","sordZ":{"bset":{"$expression":"ALL"}}}}},{"method":"deleteSord","params":{"parentId":{"$expression":"str(PARENT)"},"objId":{"$expression":"oid"},"deleteOptions":{"deleteFinally":false}}},{"method":"deleteSord","params":{"parentId":{"$expression":"str(PARENT)"},"objId":{"$expression":"oid"},"deleteOptions":{"deleteFinally":true}}}]
public final class Main {
  public static void main(String[] args) throws Exception {
    var elo = EloClient.connect(); // ELOPG_* overrides local teaching defaults.
    System.out.println(elo.call("linkSords", "{\"fromId\":{\"$expression\":\"a\"},\"toIds\":[{\"$expression\":\"b\"}],\"linkZ\":{\"bset\":\"1\"}}"));
    System.out.println(elo.call("unlinkSords", "{\"fromId\":{\"$expression\":\"a\"},\"toIds\":[{\"$expression\":\"b\"}],\"linkZ\":{\"bset\":\"1\"}}"));
    System.out.println(elo.call("createSord", "{\"parentId\":{\"$expression\":\"str(PARENT)\"},\"maskId\":0,\"editInfoZ\":{\"bset\":\"1\",\"sordZ\":{\"bset\":{\"$expression\":\"ALL\"}}}}"));
    System.out.println(elo.call("checkinSord", "{\"sord\":{\"$expression\":\"tpl\"},\"sordZ\":{\"bset\":{\"$expression\":\"ALL\"}},\"unlockZ\":{\"bset\":\"1\"}}"));
    System.out.println(elo.call("checkoutSord", "{\"objId\":{\"$expression\":\"a\"},\"editInfoZ\":{\"bset\":\"1\",\"sordZ\":{\"bset\":{\"$expression\":\"ALL\"}}}}"));
    System.out.println(elo.call("deleteSord", "{\"parentId\":{\"$expression\":\"str(PARENT)\"},\"objId\":{\"$expression\":\"oid\"},\"deleteOptions\":{\"deleteFinally\":false}}"));
    System.out.println(elo.call("deleteSord", "{\"parentId\":{\"$expression\":\"str(PARENT)\"},\"objId\":{\"$expression\":\"oid\"},\"deleteOptions\":{\"deleteFinally\":true}}"));
  }
}
